### PR TITLE
Simple refactoring of DbContext constructors

### PR DESCRIPTION
**Pull request by ajcvickers**
Refactor to keep existing null checks and increase code reuse at the expense of readonly fields.

//cc @divega

### CODE REVIEW

File: src/EntityFramework.Core/DbContext.cs [hunk 1]
```diff
@@ -24,33 +24,47 @@ public class DbContext : IDisposable, IDbContextServices
...
+            Check.NotNull(serviceProvider, "serviceProvider");
+            Check.NotNull(options, "options");
 
+            Initialize(serviceProvider, options);
+        }
+
+        private void Initialize(IServiceProvider serviceProvider, DbContextOptions options)
+        {
+            InitializeSets(serviceProvider, options);
+            _contextServices = new LazyRef<DbContextServices>(() => InitializeServices(serviceProvider, options));
```
Comment (divega) [inline src/EntityFramework.Core/DbContext.cs]: Did you consider having out parameters `contextServices`, `logger` and `setInitializer`? You can keep the fields read-only that way.

**Comment by divega** [comment]
:shipit: although you already had convinced me that this wasn't a big deal.

File: src/EntityFramework.Core/DbContext.cs (hunk 1 above)
Comment (ajcvickers) [inline src/EntityFramework.Core/DbContext.cs]: I would pretty much never consider having out parameters.

Comment (divega) [inline src/EntityFramework.Core/DbContext.cs]: LOL

**Pull request merged**